src/smt/smt_engine.cpp

```diff
@@ -1603,12 +1605,57 @@ void SmtEngine::setDefaults() {
   if(d_logic.isTheoryEnabled(THEORY_ARRAYS) || d_logic.isTheoryEnabled(THEORY_DATATYPES) || d_logic.isTheoryEnabled(THEORY_SETS)) {
     if(!d_logic.isTheoryEnabled(THEORY_UF)) {
       LogicInfo log(d_logic.getUnlockedCopy());
-      Trace("smt") << "because a theory that permits Boolean terms is enabled, also enabling UF" << endl;
+      Notice() << "Enabling UF because a theory that permits Boolean terms is "
+                  "enabled"
+               << endl;
       log.enableTheory(THEORY_UF);
       d_logic = log;
       d_logic.lock();
     }
   }
+  if (d_logic.isTheoryEnabled(THEORY_ARITH) && !d_logic.isLinear())
+  {
+    // Non-linear arithmetic requires UF to deal with division/mod because
+    // their expansion introduces UFs for the division/mod-by-zero case.
+    if (!d_logic.isTheoryEnabled(THEORY_UF))
```

Comment:
Can we introduce a flag `needsUf` and then only do this Logic setting once, also with a string for the reason?
This is similar to what I did for disabling models: https://github.com/CVC4/CVC4/blob/master/src/smt/smt_engine.cpp#L2283